controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,377 @@
+		WithWorkerMachineDeploymentClasses([]clusterv1.MachineDeploymentClass{*machineDeploymentClass, *machineDeploymentClass2}).
+		Build()
+
+	// 3) A Cluster object including a Cluster Topology Object and the MachineDeploymentTopology objects used in the
+	// Cluster Topology.
+	machineDeploymentTopology := builder.MachineDeploymentTopology("mdm").
+		WithClass(workerClassName).
+		WithReplicas(3).
+		Build()
+	machineDeploymentTopology2 := builder.MachineDeploymentTopology("mdm2").
+		WithClass(workerClassName).
+		WithReplicas(1).
+		Build()
+	cluster := builder.Cluster(ns.Name, clusterName).
+		WithTopology(
+			builder.ClusterTopology().
+				WithClass(clusterClass.Name).
+				WithMachineDeployment(machineDeploymentTopology).
+				WithMachineDeployment(machineDeploymentTopology2).
+				WithVersion("1.22.2").
+				WithControlPlaneReplicas(3).
+				Build()).
+		Build()
+
+	// Create a set of initObjects from the objects above to add to the API server when the test environment starts.
+	initObjs := []client.Object{
+		clusterClass,
+		cluster,
+		infrastructureClusterTemplate,
+		infrastructureMachineTemplate,
+		bootstrapTemplate,
+		controlPlaneTemplate,
+	}
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+
+	for _, obj := range initObjs {
+		err := env.Create(ctx, obj)
+		g.Expect(err).ToNot(HaveOccurred())
+	}
+	defer func() {
+		for _, obj := range initObjs {
+			err := env.Delete(ctx, obj)
+			g.Expect(err).ToNot(HaveOccurred())
+		}
+	}()
+	actualCluster := &clusterv1.Cluster{}
+	// Get Cluster object and check that relevant fields are set as expected.
+	g.Eventually(func() bool {
+		return assertClusterReconcile(ns.Name, clusterName, actualCluster)
+	}, timeout).Should(BeTrue())
+
+	// Check if InfrastructureCluster has been created and has the correct labels and annotations.
+	g.Eventually(func() bool {
+		return assertInfrastructureClusterReconcile(actualCluster)
+	}, timeout).Should(BeTrue())
+
+	// Check if ControlPlane has been created and has the correct version, replicas, labels and annotations.
+	g.Eventually(func() bool {
+		return assertControlPlaneReconcile(actualCluster, clusterClass)
+	}, timeout).Should(BeTrue())
+	g.Eventually(func() bool {
+		return assertMachineDeploymentsReconcile(actualCluster)
+	}, timeout).Should(BeTrue())
+}
+
+// assertClusterReconcile checks if the Cluster object:
+// 1) Has its InfrastructureReference and ControlPlane reference added correctly.
+// 2) InfrastructureReference and ControlPlaneRef have the expected Group, Version and Kind.
+func assertClusterReconcile(namespace, clusterName string, cluster *clusterv1.Cluster) bool {
+	key := client.ObjectKey{Name: clusterName, Namespace: namespace}
+	if err := env.Get(ctx, key, cluster); err != nil {
+		return false
+	}
+	// Check if relevant managed topology labels are present.
+	if ok := assertClusterTopologyOwnedLabel(cluster); !ok {
+		return false
```

Comment:
How do we know what's broken without stepping through the code? Usually we just assert (aka g.Expect) everywhere and then we get a clean error message. I'm aware this doesn't work here. 

Do we have to g.Eventually all those assertions (l.124++)? Couldn't we just retry until the reconciliation is done and then to plain assertations (the regular way with g.Expect)?

If we really have to do it, can we maybe return an error in the assert funcs and check in Eventually that it's nil? That way we get an error message after the error didn't vanish after timeout (and we also don't need logs)

Reply:
The logs are what I was using, but I can switch this to error returns for sure.

Reply:
Maybe let's see if Fabrizio aggrees (@fabriziopandini)

Reply:
I've pushed a new commit with the error flow - it makes the overall test probably easier to read I think, and it's more in line with other tests too, though there's lots of variety in the codebase.